thewalrus/_hafnian.py

```diff
@@ -55,6 +57,32 @@ def input_validation(A, rtol=1e-05, atol=1e-08):
     return True
 
 
+def bandwidth(A):
+    """Calculates the upper bandwidth of the matrix A.
+    Args:
+        A (array): input matrix
+    Returns:
+        (int): bandwidth of matrix
+    """
+    n, _ = A.shape
+    for i in range(n):
+        vali = np.diag(A, i)
+        if np.allclose(vali, 0):
+            return i - 1
+    return n - 1
+
+
+def powerset(iterable):
+    """Calculates the powerset of a list.
+    Args:
+        iterable (list): input list
+    Returns:
+        (chain): chain of all subsets of input list
+    """
+    s = list(iterable)
```

Comment:
is this necessary?

Reply:
The powerset function? Im not sure, but the way we thought of implementing the algorithm involved needing to find a powerset.

Reply:
@timjank :I think @ziofil means having to turn the `iterable` into a list.

Reply:
Ahh ok, sorry about that. In that case I have no idea lol